File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/levels/traps/LightningTrap.java

package com.watabou.pixeldungeon.levels.traps;

import com.nyrds.pixeldungeon.levels.objects.ITrigger;
import com.nyrds.pixeldungeon.mechanics.NamedEntityKind;
import com.nyrds.pixeldungeon.ml.R;
import com.nyrds.platform.util.StringsManager;
import com.watabou.noosa.Camera;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.ResultDescriptions;
import com.watabou.pixeldungeon.actors.Actor;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.effects.CellEmitter;
import com.watabou.pixeldungeon.effects.Lightning;
import com.watabou.pixeldungeon.effects.particles.SparkParticle;
import com.watabou.pixeldungeon.scenes.GameScene;
import com.watabou.pixeldungeon.utils.GLog;
import com.watabou.pixeldungeon.utils.Utils;
import com.watabou.utils.Random;

import org.jetbrains.annotations.Nullable;

public class LightningTrap implements ITrigger{

	// 00x66CCEE
	
	public static void trigger( int pos, @Nullable Char ch ) {
		if (ch == null){
			ch = Actor.findChar(pos);
		}
		if (ch != null) {
			ch.damage( Math.max( 1, Random.Int( ch.hp() / 3, 2 * ch.hp() / 3 ) ), LIGHTNING );
			if (ch == Dungeon.hero) {
				
				Camera.main.shake( 2, 0.3f );
				
				if (!ch.isAlive()) {
                    Dungeon.fail( Utils.format( ResultDescriptions.getDescription(ResultDescriptions.Reason.TRAP), StringsManager.getVar(R.string.LightningTrap_Name), Dungeon.depth ) );
                    GLog.n(StringsManager.getVar(R.string.LightningTrap_Desc));
				} else {
					ch.getBelongings().charge( false );
				}
			}
			
			int[] points = new int[2];
			
			points[0] = pos - Dungeon.level.getWidth();
			points[1] = pos + Dungeon.level.getWidth();
			GameScene.addToMobLayer( new Lightning( points) );
			
			points[0] = pos - 1;
			points[1] = pos + 1;
			GameScene.addToMobLayer( new Lightning( points) );
		}
		
		CellEmitter.center( pos ).burst( SparkParticle.FACTORY, Random.IntRange( 3, 4 ) );
		
	}
	
	public static final Electricity LIGHTNING = new Electricity();
	public static class Electricity implements NamedEntityKind {
		@Override
		public String getEntityKind() {
			return getClass().getSimpleName();
		}

		@Override
		public String name() {
			return getEntityKind();
		}
	}

	@Override
	public void doTrigger(int cell, Char ch) {
		trigger(cell,ch);
	}
}

File: scripts/traps/LightningTrap.lua
local RPD = require "scripts/lib/commonClasses"
local trap = require "scripts/lib/trap"

-- java LightningTrap: shock whoever is on the cell for a fraction of their hp,
-- shake the camera and drain wands when the hero is hit, custom death report
-- when it kills (Electricity is not a Doom source - without the explicit
-- fail the rankings cause would be lost), two crossing bolt visuals.
return trap.init(
    function (cell, char, data)
        if char == nil then
            char = RPD.Actor:findChar(cell)
        end
        if char ~= nil then
            local hp = char:hp()
            -- java Random.Int(a,b) is [a,b)
            local dmg = math.max(1, RPD.Random:Int(math.floor(hp / 3), math.floor(2 * hp / 3)))
            char:damage(dmg, luajava.new(RPD.Electricity))
            if char:getEntityKind() == "Hero" then
                RPD.shakeCamera(2, 0.3)
                if not char:isAlive() then
                    RPD.Dungeon:fail(RPD.JavaUtils:format(
                            RPD.ResultDescriptions:getDescription(RPD.ResultReason.TRAP),
                            { RPD.textById("LightningTrap_Name"), RPD.Dungeon.depth }))
                    RPD.glogn(RPD.textById("LightningTrap_Desc"))
                else
                    char:getBelongings():charge(false)
                end
            end
            RPD.Lightning:spawnBolts(cell)
        end

        RPD.Sfx.CellEmitter:center(cell):burst(RPD.Sfx.SparkParticle.FACTORY, RPD.Random:IntRange(3, 4))
    end
)
